File: 2359-find-closest-node-to-given-two-nodes/2359-find-closest-node-to-given-two-nodes.go
func closestMeetingNode(edges []int, node1 int, node2 int) int {
    return bellmanFord(edges, node1,node2)
}
func bellmanFord(edges []int, node1 int, node2 int) int {
    bf1 := calculateBellmanFord(node1, edges)
    bf2 := calculateBellmanFord(node2,edges)
    
	//max Integer
    minPath := 1 << 31
    result := -1
    for i:=range bf1 {
        if bf1[i] == 1 << 31 || bf2[i] == 1 << 31{
            continue
        }
        
        v := max(bf1[i],bf2[i])
        if v < minPath{
            minPath = v
            result = i
        }
    }
    
    return result
}

func calculateBellmanFord(root int, edges []int) []int{
    distances := make([]int, len(edges))
    for i:=range distances {
        distances[i]=1 << 31
    }
    distances[root]=0
    
    node := root
    for edges[node] != -1 {
        next := edges[node]
        if distances[next] > distances[node] + 1 {
            distances[next]=distances[node] + 1
            node = next
        }else{
            break
        }
    }
    
    return distances
}

func max(a,b int) int {
    if a > b {
        return a
    }
    
    return b
}

File: 2359-find-closest-node-to-given-two-nodes/2359-find-closest-node-to-given-two-nodes.cs
public class Solution {
    public int ClosestMeetingNode(int[] edges, int node1, int node2) {
        Dictionary<int, int> d1 = new (), d2 = new ();

        int iterIndex = node1, distance = 0;
        while (iterIndex != -1 && !d1.ContainsKey(iterIndex))
        {
            d1[iterIndex] = distance++;
            iterIndex = edges[iterIndex];
        }

        int ansIndex = int.MaxValue, ansDistance = int.MaxValue;
        iterIndex = node2;
        distance = 0;
        while (iterIndex != -1 && !d2.ContainsKey(iterIndex))
        {
            d2[iterIndex] = distance++;

            if (d1.ContainsKey(iterIndex))
            {
                int m = Math.Max(d1[iterIndex], d2[iterIndex]);
                if (m < ansDistance || (m == ansDistance && iterIndex < ansIndex))
                {
                    ansIndex = iterIndex;
                    ansDistance = m;
                }
            }
            iterIndex = edges[iterIndex];
        }

        if (ansIndex == int.MaxValue)
            return -1;
        return ansIndex;
    }
}